Add post publish QS task

planarvoid opened the pull request:
Fixes #13845

This PR adds logic to show the quick start focus point on the FAB from the improved my site fragment. I was considering moving the FAB to the fragment (because I think it logically belongs there) but the changes I'd have to make were massive and the code with two access point gets really complicated. In the end the code shows the FAB from the fragment but handles the logic to hide as before - through the view model which now calls QuickStartRepository when the flag is enabled.

![publish_post_video](https://user-images.githubusercontent.com/1079756/106489841-7cc3d880-64b5-11eb-96da-18470dd43516.gif)


To test:
- Turn on the MySiteImprovements flag and restart the app
- Make sure you are using a new site with Quick Start in progress
- Click on "Publish post" QS card
- Notice a Snackbar with a correct message
- Notice a focus point on the FAB
- Click on another QS card ("Update title")
- Notice the focus point is gone
- Click on the "Publish post" QS card again
- Click on the FAB
- Notice the focus point on the "Blog post" item
- Click on the "Blog post" item 
- Go back
- Notice the "Publish post" card is now gone

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

planarvoid committed:
Add logic to show FAB focus point from the ImprovedMySiteFragment.kt

planarvoid committed:
Merge branch 'develop' into feature/13845-add-post-publish-qs-task

renanferrari reviewed (commented):
Thanks for the changes! Code looks pretty good and I've tested the app and everything seems to work fine, except for the scenario in the video below:

https://user-images.githubusercontent.com/830056/106543906-6897cf80-64e5-11eb-9720-9f19d46096ac.mp4

While working on #13846, I also came across similar issues myself as that task requires us to show a focus point on the navigation bar. I decided to try a different approach there, by treating that focus point as an event rather than a state and it seemed to work well. 

I've made [a draft of that implementation available here](https://github.com/wordpress-mobile/WordPress-Android/pull/13934), so you can take a look. If you think it's worth it, we can try following the same approach here. If not, we could try something else to achieve some consistency between the implementations. Either way, let me know what you think.

planarvoid committed:
Merge branch 'develop' into feature/13845-add-post-publish-qs-task

planarvoid committed:
Move FAB focus point logic to Main activity and view model

planarvoid commented:
Great catch @renanferrari . I've looked into it and I've actually implemented the logic differently so it makes more sense. I've added the `focusPoint` field to the FAB view model. Since it's already hidden/shown based on the live data, I've just added one additional source and that's the active task. The change of the active task now redraws the FAB and hides/shows it. I'm much happier with this solution :-). Her's how it works now:

![publish_post_fixed](https://user-images.githubusercontent.com/1079756/106608168-5b6df580-6564-11eb-9c92-c713859c852a.gif)

planarvoid committed:
Remove unused imports

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java:
```diff
@@ -420,13 +441,31 @@ private void initViewModel() {
+                } else if (!fabUiState.isFocusPointVisible()) {
+                    mHandler.removeCallbacks(mShowFabFocusPoint);
+                    mHandler.post(() -> {
+                        View focusPointView =
+                                findViewById(R.id.fab_container).findViewById(R.id.quick_start_focus_point);
+                        if (focusPointView != null) {
+                            ViewParent directParent = focusPointView.getParent();
+                            if (directParent instanceof ViewGroup) {
+                                ((ViewGroup) directParent).removeView(focusPointView);
+                            }
+                        }
```
It seems that we could replace this with `QuickStartUtils::removeQuickStartFocusPoint`?

planarvoid replied:
I've previously used that method but the problem I had there is that it wasn't removing the correct focus point. For example when you clicked on "Publish post" and then on "Update site title", the site title focus point was removed because that method just finds any view in the parent with that ID and it was basically a race condition.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/ImprovedMySiteFragment.kt:
```diff
@@ -173,7 +173,9 @@ class ImprovedMySiteFragment : Fragment(),
                 loadGravatar(uiModel.accountAvatarUrl)
                 when (val state = uiModel.state) {
-                    is State.SiteSelected -> loadData(state.items)
+                    is State.SiteSelected -> {
+                        loadData(state.items)
+                    }
```
Big nitpick here, just to maintain consistency.

```suggestion
                    is State.SiteSelected -> loadData(state.items)
```

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/QuickStartRepository.kt:
```diff
@@ -60,9 +63,10 @@ class QuickStartRepository
+    private val _activeTask = MutableLiveData<QuickStartTask?>()
     private val _onSnackbar = MutableLiveData<Event<SnackbarMessageHolder>>()
     val onSnackbar = _onSnackbar as LiveData<Event<SnackbarMessageHolder>>
+    val activeTask = _activeTask as LiveData<QuickStartTask?>
```
Good call on making this public! 👍

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/QuickStartRepository.kt:
```diff
@@ -74,14 +78,14 @@ class QuickStartRepository
         emit(QuickStartUpdate())
         refresh.asFlow().map {
-            if (quickStartUtils.isQuickStartInProgress(siteId)) {
+            if (quickStartUtils.isQuickStartInProgress(siteId) || true) {
```
This seem to have been left here by mistake?

```suggestion
            if (quickStartUtils.isQuickStartInProgress(siteId)) {
```

planarvoid replied:
yeah, good catch! I use it for testing :-)

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/QuickStartRepository.kt:
```diff
@@ -143,14 +147,18 @@ class QuickStartRepository
         job.cancel()
     }
 
+    fun shouldShowPublishPostQuickStartTask(): Boolean {
+        return activeTask.value == PUBLISH_POST
+    }
```
Since `activeTask` is now public, I wonder if we could remove this?

renanferrari reviewed (approved):
Thanks for the changes! I can confirm the app now works as expected and I really like the solution you found as well :)

I left a few comments below, but they are mostly nitpicks. Feel free to merge this after addressing them and the errors in the tests.

planarvoid committed:
Fix a test

planarvoid committed:
Update WordPress/src/main/java/org/wordpress/android/ui/mysite/ImprovedMySiteFragment.kt

Co-authored-by: Renan Ferrari <renan.fb@gmail.com>

planarvoid committed:
Update WordPress/src/main/java/org/wordpress/android/ui/mysite/QuickStartRepository.kt

Co-authored-by: Renan Ferrari <renan.fb@gmail.com>

planarvoid committed:
Remove unnecessary method

planarvoid committed:
Merge branch 'feature/13845-add-post-publish-qs-task' of github.com:wordpress-mobile/WordPress-Android into feature/13845-add-post-publish-qs-task

renanferrari reviewed (approved):
LGTM! Feel free to merge once the lint errors are fixed.

planarvoid committed:
Fix ktlint issue

planarvoid committed:
Merge branch 'develop' into feature/13845-add-post-publish-qs-task
